rust-symcrypt/src/hmac.rs

```diff
@@ -1,190 +1,436 @@
             symcrypt_sys::SymCryptHmacSha256Append(
-                &mut self.state,
+                &mut self.inner.state,
                 data.as_ptr(),
                 data.len() as symcrypt_sys::SIZE_T,
             )
         }
     }
 
-    pub fn result(&mut self, result: &mut [u8; SHA256_HMAC_RESULT_SIZE]) {
+    fn result(mut self) -> Self::Result {
+        let mut result = [0u8; SHA256_HMAC_RESULT_SIZE];
         unsafe {
-            symcrypt_sys::SymCryptHmacSha256Result(&mut self.state, result.as_mut_ptr());
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptHmacSha256Result(&mut self.inner.state, result.as_mut_ptr());
         }
+        result
     }
 }
 
-impl Drop for HmacSha256State {
-    fn drop(&mut self) {
+/// Creates a clone of the current HmacSha256State. Clone will create a new state field but will reference the same
+/// expanded_key of the current HmacSha256State; therefore increasing the refcount on expanded_key field.
+impl Clone for HmacSha256State {
+    fn clone(&self) -> Self {
+        let mut new_state = HmacSha256State {
+            inner: Box::pin(HmacSha256Inner {
+                state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE::default(),
+                expanded_key: Pin::clone(&self.inner.expanded_key),
+            }),
+        };
         unsafe {
-            symcrypt_sys::SymCryptWipe(
-                ptr::addr_of_mut!(self.state) as *mut c_void,
-                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptHmacSha256StateCopy(
+                &self.inner.state,
+                &self.inner.expanded_key.0,
+                &mut new_state.inner.state,
             );
+        }
+        new_state
+    }
+}
 
+impl Drop for HmacSha256State {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
             symcrypt_sys::SymCryptWipe(
-                ptr::addr_of_mut!(self.expanded_key) as *mut c_void,
-                mem::size_of_val(&mut self.expanded_key) as symcrypt_sys::SIZE_T,
+                ptr::addr_of_mut!(self.inner.state) as *mut c_void,
+                mem::size_of_val(&mut self.inner.state) as symcrypt_sys::SIZE_T,
             );
         }
     }
 }
 
-pub fn hmac_sha256(key: &[u8], data: &[u8], result: &mut [u8; SHA256_HMAC_RESULT_SIZE]) {
+/// Stateless Hmac functions for Sha256, using Result<> here for more friendly rust
+pub fn hmac_sha256(
+    key: &[u8],
+    data: &[u8],
+) -> Result<[u8; SHA256_HMAC_RESULT_SIZE], SymCryptError> {
+    let mut result = [0u8; SHA256_HMAC_RESULT_SIZE];
     unsafe {
-        let mut expanded_key = symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default();
-        symcrypt_sys::SymCryptHmacSha256ExpandKey(
-            &mut expanded_key,
+        // SAFETY: FFI calls
+        let mut expanded_key = Arc::new(HmacSha256ExpandedKey(
+            symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default(),
```

Comment:
This expanded_key won't  need to be shared. Don't need the Arc::new()

= HmacSha256ExpandedKey(symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default(),);